### DISCUSSION TITLE
Digital noise reduction and removal (after analysis)

### DISCUSSION THREAD

**Discussion by phillprice** [mcguirepr89/BirdNET-Pi#597] [Ideas]
My microphone is very noisy. While I don't believe its causing problems with identification its certainly annoying when listening back.

I notice that sox can help do some noise reduction, if you have a sample with just noise in it. Guess what - I have many hours a day with just noise in it :)

https://unix.stackexchange.com/a/427343 is where I sourced the commands, although I've not automated them yet, and I think a backup file would be useful, just in case if you were to automate this.

https://soundcloud.com/phillprice/sets/tawny-owls?si=1f58816152ff4528970e13c486d484cb

Basic steps are 

Install
```
sudo apt install sox libsox-fmt-*
```

Find a file that is birdless (from overnight / now) and copying it to home
```
cp ~/BirdSongs/October-2022/08-Saturday/2022-10-08-birdnet-14\:51\:20.wav ~
```

Make a noise profile (mine complained `sox WARN wav: Premature EOF on .wav input file`)
```
sox 2022-10-08-birdnet-14\:51\:20.wav -n noiseprof noise_profile_file
```

Grab a file to home to test it (TEST ON THE COPY)
```
cp ~/BirdSongs/Extracted/By_Date/2022-10-08/Tawny_Owl/Tawny_Owl-91-2022-10-08-birdnet-02:22:44.mp3 ~
```

Test the filter
```
sox Tawny_Owl/Tawny_Owl-91-2022-10-08-birdnet-02\:22\:44.mp3 Tawny_Owl/Tawny_Owl-91-2022-10-08-birdnet-02\:22\:44-filtered.mp3 noisered noise_profile_file 0.31
```

Tweak and repeat at will. PLEASE **DON'T DO THIS ON ORIGINALS**

**Comment by matthew73210** [comment]
Thats actually a pretty neat idea, this could probs be done on samples before anaylis? Every x minutes take an empty samble and use as a filter. @mcguirepr89 what do you think? Maybe have it as an option for noisy environments or ones where there's always interference.

**Reply by ehpersonal38** [reply]
There's a way to do this automatically with pulseaudio by applying an active noise cancellation filter to your microphone. I've done it before and it works a charm! Could be something to look into if more people see the value of it

**Reply by peflo** [reply]
The pulseaudio cancelling does work nicely - I'm surrounded by busy roads and in earshot of a local freeway so I have  a steady background roar of traffic.  Enabling pulseaudio's echo-cancel module (based on the instructions here: https://askubuntu.com/questions/18958/realtime-noise-removal-with-pulseaudio) cut the audible noise on the stream down dramatically and seems to be giving a better success rate on bird calls it was missing before.  I might also take a stab at enabling a high-pass filter to see if that helps.  I should also add that I'm not uploading these anywhere - I assume this level of manipulation is frowned on for official use.

**Reply by phillprice** [reply]
Which of the answers did you follow?

**Reply by rgbdeluxe** [reply]
> There's a way to do this automatically with pulseaudio by applying an active noise cancellation filter to your microphone. I've done it before and it works a charm! Could be something to look into if more people see the value of it

I would be interested in having this kind of option. my set up is also at a busy street and would probably benefit from a noise removal.

**Reply by ehpersonal38** [reply]
Basically the steps are
1. Run this command to list your audio devices: `LANG=C pacmd list-sources | grep name: | grep input`. Find your input device (your microphone) and copy its name. You only want to copy the string in between the two < > symbols. For example, I would copy:
`alsa_input.usb-C-Media_Electronics_Inc._USB_Audio_Device-00.mono-fallback`
**NOT**
`name: <alsa_input.usb-C-Media_Electronics_Inc._USB_Audio_Device-00.mono-fallback>`
2. Open the file `/etc/pulse/default.pa` in any text editor. Make sure to use sudo otherwise you won't have permissions to write.
3. Add the following code to the very end of the file, replacing the input device name that comes after `source_master`, with your own from step 1.
```
.ifexists module-echo-cancel.so
load-module module-echo-cancel source_master=alsa_input.usb-C-Media_Electronics_Inc._USB_Audio_Device-00.mono-fallback aec_method=webrtc source_name=echocancel sink_name=echocancel1
set-default-source echocancel
set-default-sink echocancel1
.endif
```
4. Run this command to reload pulseaudio: `pulseaudio -k`

Now your microphone should automatically cancel noise! This will persist across reboots.

**Reply by peflo** [reply]
What I did was based on the accepted answer - I didn't take notes on the exact process and I don't know if this is the optimal way to do it, but in a nutshell here's what I did - you should be able to do all of this through the web console:

1. Edit `/etc/pulse/default.pa` to add `load-module module-echo-cancel` in the "Modules to allow autoloading of filters" block
2. Run `sudo pulseaudio -k` and `sudo pulseaudio --start` to restart pulseaudio and pick up the change
3. Run `pacmd list-sources | grep -e 'index:' -e device.string -e 'name:'` to list the audio device names. We're looking for one that ends with `echo-cancel`, mine was `alsa_input.usb-0c76_USB_Audio_Device-00.mono-fallback.echo-cancel`.  
4. Edit `/etc/pulse/default.pa` again, and add 'set-default-source <device name from step 3>
5. Reboot the pi, and enjoy the quieter stream!  
 
I haven't tried yet, but I assume that reverting the default.pa file and rebooting would switch things back to normal.

**Reply by peflo** [reply]
@ehpersonal38 - Did you need to reboot or restart birdnet after restarting pulseaudio?   The livestream stopped working for me at that point so I assumed that restarting the service effectively broke the connection to the device.

**Reply by ehpersonal38** [reply]
Interesting! I think both our methods took slightly different steps to get to same result.

I didn't need to restart any services or anything after reloading pulse. The livestream did cut out, but after a browser refresh it was back up and running.

Side note I just remembered - when I reloaded pulseaudio after enabling the echo cancelation module, it turned my microphone gain way down to like 10, so I had to go into alsamixer and put the gain back up.

**Reply by blogdemoi** [reply]
> Basically the steps are
> 
>     1. Run this command to list your audio devices: `LANG=C pacmd list-sources | grep name: | grep input`. Find your input device (your microphone) and copy its name. You only want to copy the string in between the two < > symbols. For example, I would copy:
>        `alsa_input.usb-C-Media_Electronics_Inc._USB_Audio_Device-00.mono-fallback`
>        **NOT**
>        `name: <alsa_input.usb-C-Media_Electronics_Inc._USB_Audio_Device-00.mono-fallback>`

Hello,

I tried your command and it will only list: 

> name: <alsa_input.usb-046d_HD_Pro_Webcam_C920_96B3555F-02.analog-stereo>

which in my case is the microphone of the usb camera taking pictures of my garden. However my microphone recording the birds is a simple lavalier mic (Boya) plugged in a usb soudcard, then plugged in my LePotato and it doesn't appear in the list.

Any idea why?

Thank you.

**Reply by blogdemoi** [reply]
Ah sorry. I just realized that I plugged in my usb sound card but didn't reboot so the soudcard wasn't listed and the command `LANG=C pacmd list-sources | grep name: | grep input` only listed the mic of my usb webcam (that I use for the pictures of my garden).

Once I rebooted my new usb soundcard and its lavalier mic are now recognized:

> LANG=C pacmd list-sources | grep name: | grep input`
> name: <alsa_input.usb-KTMicro_KT_USB_Audio_2021-06-07-0000-0000-0000--00.mono-fallback>
> name: <alsa_input.usb-046d_HD_Pro_Webcam_C920_96B3555F-02.analog-stereo>

However I now have a new problem. It seems impossible to select the usb soundcard as the default audio source in alsamixer.

![sParH3W](https://user-images.githubusercontent.com/83965802/227384831-7c34eb2f-7431-439b-b982-ba66cb5a3eb2.png)

![60Gr9cC](https://user-images.githubusercontent.com/83965802/227384841-8cb26d19-6dc4-4af4-95ae-fbe40d56a6e5.png)

Whatever I do, I still hear the sound from the webcam and not from the usb soundcard...

Edit: I unpluged the webcam, rebooted just with usb soundcard and it picked up the usb soundcard as the default audio source.

**Reply by ehpersonal38** [reply]
So is all well?

You can set the default using this guide, but I wouldn't fumble around with it unless you need to: https://askubuntu.com/a/435197

**Reply by blogdemoi** [reply]
Yes, it is ok now, I learnt a few things, namely how to add a default source for pulseaudio in /etc/pulse/default.pa with `set-default-source `.

Each time I plugged in (live) my webcam (which also has a built in mic) it became the default audio source (i don't know why) (however I want to keep my lavalier mic/usb soundcard the default). Setting the default source in /etc/pulse/default.pa solved my problem, thank you.

**Reply by Vollpflock** [reply]
Does this still work for you? My Pi is stuck after reboot, had to go back. And btw, don't you struggle with low audio volume?

**Reply by phillprice** [reply]
Yes still working although the commmand is in from as it somehow gets frequently forgotten and the noise returns. Sound seems to be fine enough to detect birds.

**Reply by peflo** [reply]
Still working for me as well - the only problem I have is that the live audio stream generally doesn't play for me now, but detection works just fine and recordings have the noise-cancellation enabled and play properly.  No issues with volume.

**Reply by harrymnel** [reply]
My microphone gain also dropped and I can barely hear it. How do I use alsamixer to put the gain back up and save it. @ehpersonal38

**Reply by Vollpflock** [reply]
https://github.com/mcguirepr89/BirdNET-Pi/wiki/Adjusting-your-sound-card

**Comment by morrowwm** [comment]
Discussion #253 is converging on this topic.
